### ISSUE TITLE
refactor polygonal/polyhedral layout

### ISSUE THREAD

**Issue by xjrc** [LLNL/conduit#504]
In order to improve interoperability with client codes, the data layout for Conduit-Blueprint's polygonal/polyhdral elements needs to be updated. Specifically, the layout needs to be changed from an in-line, stream-based sequence to an out-of-line, offset-based collection of sequences. This can be accomplished by adopting Conduit's new one-to-many schema (#511) to extract the polygon sizes from the index stream.

#### Polygonal Variables ####

- `eI`: number of vertices in 2D element I
- `eIvJ`: the index of vertex J in element I

#### Current Layout ####

```
topology:
  coordset: coords
  type: unstructured
  elements:
    shape: polygonal
    connectivity: [e1, e1v1, ..., e1ve1, e2, e2v1, ..., e2ve2, ...]
```

#### New Layout ####

```
topology:
  coordset: coords
  type: unstructured
  elements:
    shape: polygonal
    connectivity: [e1v1, ..., e1ve1, e2v1, ..., e2ve2, ...]
    sizes: [e1, e2, e3, ...]
    offsets: [0, e1, e1+e2, ...]
```

This situation changes slightly in 3D because the top-level, out-of-line `begins` sequence does not fully describe the internal structure of an arbitrary element (e.g. how the faces that comprise the 3D element are laid out). For such cases, an extra offset-based sequence needs to be generated.

#### Polygonal Variables ####

- `eI`: number of sub-elements in 3D element I
- `eIsJ`: the index of sub-element (face) J of 3D element I
- `sIvJ`: the index of vertex J in sub-element (face) I 
- `|eIsJ|`: number of vertices in sub-element (face) J of 3D element I
- `eI_sJ_vK`: the index of vertex K in sub-element J in element I

#### Current Layout ####

```
topology:
  coordset: coords
  type: unstructured
  elements:
    shape: polyhedral
    connectivity: [e1, |e1_s1|, e1_s1_v1, ..., e1_s1_vs1, ..., |e1_se1|, e1_se1_v1,  ..., e1_se1_vse1, e2, ...]
```

#### New Layout ####

```
topology:
  coordset: coords
  type: unstructured
  elements:
    shape: polyhedral
    connectivity: [e1s1, ..., e1se1, e2s1, ..., e2se2, ...]
    sizes: [e1, e2, e3, ...]
    offsets: [0, e1, e1+e2, ...]
  subelements:
    shape: polygonal
    connectivity: [s1v1, ..., s1vs1, s2vs1, ..., s2vs2, ...]
    sizes: [s1, s2, s3, ...]
    offsets: [0, s1, s1+s2, ...]
```

In order to help illustrate the meaning of this refactoring, here are the [Blueprint polygonal/polyhedral examples](https://llnl-conduit.readthedocs.io/en/latest/blueprint_mesh.html#polygonal-polyhedral-topologies) in the new format:

#### Example Diagram ####

```
       4-----5
       |`\   |
 e1 -> |  \  | <- e0
       |   \.|
       7-----6
```

#### New Layout ####

```
topology:
  coordset: coords
  type: unstructured
  elements:
    shape: polygonal
    connectivity: [4, 6, 5, 7, 6, 4]
    sizes: [3, 3]
    offsets: [0, 3]
```

#### Example Diagram ####

```
         0
        /|\
       / | \ <- e0
      /  |  \
     /_.-3-._\
    1.,  |  ,.4
     \ `'2'` /
      \  |  /
 e1 -> \ | /
        \|/
         5
```

#### New Layout ####

```
topology:
  coordset: coords
  type: unstructured
  elements:
    shape: polyhedral
    connectivity: [0, 1, 2, 3, 4, 0, 5, 6, 7, 8]
    sizes: [5, 5]
    offsets: [0, 5]
  subelements:
    shape: polygonal
    connectivity: [1, 2, 4, 3, 0, 1, 2, 0, 2, 4, 0, 3, 4, 0, 1, 3, 5, 1, 2, 5, 2, 4, 5, 3, 4, 5, 1, 3]
    sizes: [4, 3, 3, 3, 3, 3, 3, 3, 3]
    offsets: [0, 4, 7, 10, 13, 16, 19, 22, 25, 28]
```

**Comment by nselliott** [comment]
I'm trying to wrap my head around the 3D proposal.  The old format had elements, sub-elements, and vertices, while the new one introduces faces.  Are faces the same as sub-elements?  Also, referring to both "element I" and "face I" is confusing.  I'm not sure how the fI values relate to the other arrays.

  I might be confused because of the use of the symbol I to identify both an element and a face.  Is there a difference between a face and a sub-element?

**Comment by xjrc** [comment]
@nselliott: I used the "face" terminology in order to help me think through the new layout, but they are actually the same as "sub-elements" from before. As with the old layout, the new 3D layout simply enumerates the constituent polygonal faces (sub-elements) of a polyhedron and the constituent vertices of those faces (sub-elements); it now just uses two arrays to do so instead of one. Lamentably, this does make things a bit more confusing since it requires multiple levels of indirection, but we're brainstorming strategies to improve this.

To be more expressly clear about how this may work, here's how the polyhedral example from the Blueprint documentation would look using this new polyhedral structure:

```
// Example Diagram:
//
//         0
//        /|\
//       / | \ <- e0
//      /  |  \
//     /_.-3-._\
//    1.,  |  ,.4
//     \ `'2'` /
//      \  |  /
// e1 -> \ | /
//        \|/
//         5
//

int64 polyhedral_elem_begins[] = {0, 5, 10};
//                                   ^   ^
//                                   |   +--------------+
//            elem 0 subelem indices +---+              | elem 1 subelem indices
//                                       |              |
//                                 |-----------|  |-----------|
int64 polyhedral_elem_indices[] = {0, 1, 2, 3, 4, 0, 5, 6, 7, 8};
//                                 ^     ^     ^     ^     ^
//   'elem_indices' indexes into   +-+   +-+   +---+ +-+   +---+
//   'subelem_begins' array!         |     |       |   |       |
//                                   v     v       v   v       v
int64 polyhedral_subelem_begins[] = {0, 4, 7, 10, 13, 16, 19, 22, 25, 28};
//                                      ^  ^   ^
//                                      |  |   +--------------+
//                                      ++ +---------+        |
//                                       |           |        |
//                                    |--------|  |-----|  |-----|
int64 polyhedral_subelem_indices[] = {1, 2, 4, 3, 0, 1, 2, 0, 2, 4, 0, 3, 4, 0, 1, 3, 5, 1, 2, 5, 2, 4, 5, 3, 4, 5, 1, 3}
```

I hope that this helps, but please let me know if you have any more questions/suggestions; this is still very much a work in progress and we're hoping to get a minimal solution with improved indexing that works for all parties involved!

**Comment by nselliott** [comment]
@xjrc Thanks, this does clarify things a bit.  Is the final entry in the first array `{0, 5, 10}` needed to signal the length of the next array?  At first I thought that array should be `{0, 5}` since there are only two elements in the example, but I can see how one more number might be necessary.  If I'm interpreting this correctly, that should be part of the documentation.

I like that this format describes the shared sub-elements only once.  There were redundant descriptions in the old format, if I understood it correctly.

Perhaps the final array could be named `polyhedral_subelem_vertices`, as each number identifies a unique vertex?

The indirection does introduce a new learning curve, but I think once users grasp the concepts it will be easier to write client code that consumes this data in this format than with the single huge `poly_data` array.

**Comment by xjrc** [comment]
@nselliott 

>  At first I thought that array should be {0, 5} since there are only two elements in the example, but I can see how one more number might be necessary.

You are correct in identifying that there's a bit of redundant data in this specification (though, it's the 0 that could be assumed), but we keep it around to make calculations more straightforward. Suppose we want to iterate through all of polyhedral element `i`'s indices. With the current scheme, we always know the length is `begins[i+1]-begins[i]` and that our [start, end) interval is [`begins[i]`, `begins[i+1]`). Thus, we can implement our iteration like the following:

```
int64 i = /* index of element i */;
for( int64 j = polyhedral_elem_begins[i]; j < polyhedral_elem_begins[i+1]; j++ ) {
    int64 i_elem_subelem_index = polyhedral_elem_indices[j];
    // process on 'i_elem_subelem_index' //
}
```

>  If I'm interpreting this correctly, that should be part of the documentation.

Agreed; we will be sure to very clearly outline the meaning of each array and the motivation for their existence/structure.

> I like that this format describes the shared sub-elements only once. There were redundant descriptions in the old format, if I understood it correctly.

The old representation did indeed have redundant sub-elements specifications, but I'm not certain that the new method will always eliminate these redundancies. I eliminated the redundant face in my example in order to reduce the example's complexity, but I don't believe that this will be a requirement of the new representation. More than likely, we'll allow for both styles of specification (for those that may want a uniquely oriented face for each cell) and have our derivation functions generate non-redundant versions like my example.

> Perhaps the final array could be named polyhedral_subelem_vertices, as each number identifies a unique vertex?

Certainly; calling the final array `polyhedral_subelem_vertices` would be more descriptive.

> The indirection does introduce a new learning curve, but I think once users grasp the concepts it will be easier to write client code that consumes this data in this format than with the single huge poly_data array.

Agreed, and that's the main motivation for the change! Also, we have a few clients that consume data in precisely this format, so being able to support transferring their data without transforming it is an added bonus.

**Comment by cyrush** [comment]
instead of one array where it is implied that the next starting index of the polygon is +1 from the end of the current, I think using an explicit length is more flexible: 

so -  poly_begins, would become something like:
 
* offsets
* lengths (or sizes)

The offsets and lengths (or sizes) is also a general construct we can use in other cases where we have a one-to-many relationship. 


A similar case is heterogeneous element types, however there the length is not sufficient to identify an element type (tets vs quads is ambiguous w/o more info), here we might argument with another array:
 
* offsets
* lengths (or sizes)
* element_shape_ids

**Comment by xjrc** [comment]
I've updated this issue so that it's compliant with the updates outlined in #511.

**Comment by cyrush** [comment]
@xjrc looks good, I like `subelements` for the polyhedral case!

**Comment by nselliott** [comment]
@xjrc In 3D, should there be any kind of explicit mapping between elements and subelements?  I see how it's logically well-defined--if a subelement's vertices are a subset of an element's vertices, then that subelement must be a face of the element.  But if I were given an element, I'm not sure how I would find all of its faces other than by a linear search of the subelement connectivity array.

Also, are the offsets arrays optional?  In all examples shown, it looks like `offsets[n]` is always the sum of the first `n-1` entries of `sizes`.  Would there be a case where that is not true?

**Comment by cyrush** [comment]
@nselliott in the polyhedral case, we use a one-to-many relationship from each polyhedron to a set of subelements. The polyhedrons are still defined as set of faces (those are the subelements), which then are defined in terms of the vertices. 

We talked about offsets being optional, but w/o the offsets you can't random access each element (polygon or polyhedron). It can be generated easily, but then it is hard to amortized that generation, so we thought it would be best to require it.

**Comment by nselliott** [comment]
@cyrush Thanks, I was confusing myself and reading the element connectivity array as containing vertices.  It suddenly all makes sense now.

Fine with me to have offsets required.

**Comment by brunner6** [comment]
@cyrush and @xjrc : Looks good from my quick read.  A couple of things I noticed:

We noticed a significant code speedup when we switched from 64 to 32 bit ints. We use 64 for global IDs only.  We can dig up the data if you'd like.

Using the word "sub-element" doesn't feel quite right.  We have maps from zones to nodes, and nodes to zones?  Are zones sub-elements of nodes?

What if you just wanted a connectivity map? The case I'm thinking of is that you might want to define in one place what nodes are on a face.  And then say which faces are in a zone, or which faces connect to an edge.  You wouldn't want to redefine the faces->node connectivity all the time.
